### PR TITLE
Feature/ui improvements

### PR DESCRIPTION
**Pull request by mariofv**
What's new:
- Revamped `ComponentTransform2D`, that includes anchors, pivots, ...
- Revamped UI render process. It uses now `ComponentCanvas` and `ComponentCanvasRenderer`.
- Revamped `ComponentText`. `Font` is now a resource and can be changed. Also, alignment can be set.
- `ComponentCanvas` uses viewport size, and render different in game camera and scene camera.

What's missing:
- `ComponentText` is a bit buggy a can crash we certain size combinations, needs to be improved.
- `ComponentTransform2D` anchored preset cannot be changed in the inspector. I will change it in the future with nice images! 😄 
- `ComponentButton` still not detecting clicks, I'm waiting for the EventSystem to be done.

### CODE REVIEW

File: Engine/Module/ModuleUI.h [hunk 1]
```diff
@@ -2,36 +2,21 @@
...
 #include <GL/glew.h>
-// FreeType
-#include <ft2build.h>
-#include FT_FREETYPE_H
-
-#define MAX_NUM_LAYERS 8
 
+class Component;
 class ComponentCanvas;
+class ComponentCanvasRenderer;
```
Comment (JorxSS) [inline Engine/Module/ModuleUI.h:15]: Some CS here (alphabetical order) 😈

File: Engine/Module/ModuleUI.cpp [hunk 2]
```diff
@@ -41,167 +37,81 @@ bool ModuleUI::CleanUp()
...
-	std::sort(ordered_ui.begin(), ordered_ui.end(), [](ComponentUI* left, ComponentUI* right)
+	main_canvas = nullptr;
+
+	for (auto& canvas : canvases)
 	{
-		return left->layer < right->layer;
+		if (canvas->IsEnabled())
+		{
+			main_canvas = canvas;
+			return;
```
Comment (JorxSS) [inline Engine/Module/ModuleUI.cpp:114]: Here you are just getting the first element always? We can have now more than one canvas?

File: Engine/Component/Component.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	case Component::ComponentType::UI_BUTTON:
+		return "UI Button";
+
+	case Component::ComponentType::UI_IMAGE:
+		return "UI Image";
+
+	case Component::ComponentType::UI_TEXT:
+		return "UI Text";
+	}
+}
```
Comment (Dartemiss) [inline Engine/Component/Component.cpp:69]: Why is this needed? when it is used?

File: Engine/Component/Component.h [hunk 4]
```diff
@@ -12,15 +12,21 @@ class Component
...
+		MESH_RENDERER = 2,
+		TRANSFORM = 3,
+		TRANSFORM2D = 7,
+		SCRIPT = 6,
+
+		CANVAS_RENDERER = 10,
+		UI_BUTTON = 11,
+		UI_IMAGE = 12,
+		UI_TEXT = 13,
+
```
Comment (Dartemiss) [inline Engine/Component/Component.h]: Talk with @yalania and @marcorod94  to decide the numerology of the components

File: Engine/Component/ComponentCanvas.cpp [hunk 5]
```diff
@@ -1,38 +1,127 @@
...
 }
 
 void ComponentCanvas::Delete()
 {
-	ComponentUI::Delete();
+	App->ui->RemoveComponentCanvas(this);
 }
 
-void ComponentCanvas::Save(Config& config) const
+void ComponentCanvas::Render(bool scene_mode)
```
Comment (JorxSS) [inline Engine/Component/ComponentCanvas.cpp:59]: what is the difference between scene_mode true and false?

File: Engine/Module/ModuleUI.cpp [hunk 6]
```diff
@@ -41,167 +37,81 @@ bool ModuleUI::CleanUp()
 	return true;
 }
 
-void ModuleUI::Render(const ComponentCamera* camera)
+void ModuleUI::Render(const ComponentCamera* camera, bool scene_mode)
```
Comment (JorxSS) [inline Engine/Module/ModuleUI.cpp]: Here now you don't use anymore the ComponentCamera?

File: Engine/Module/ModuleEditor.cpp [hunk 7]
```diff
@@ -129,7 +129,7 @@ update_status ModuleEditor::Update()
 		inital_scene_loaded = true;
 	}
 	//ImGui::ShowStyleEditor();
-	//ImGui::ShowDemoWindow();
+	ImGui::ShowDemoWindow();
```
Comment (JorxSS) [inline Engine/Module/ModuleEditor.cpp]: remember to comment this, or maybe we could put it as option in the main bar like Debug, Config, AI, Console. etc..

File: Engine/Module/ModuleAudio.cpp [hunk 8]
```diff
@@ -43,7 +43,7 @@ bool ModuleAudio::Init()
 		APP_LOG_ERROR("Could not initialize the Sound Engine.");
 		return false;
 	}
-	init_sound_bank = SoundManager::Init();
+	//init_sound_bank = SoundManager::Init();
```
Comment (JorxSS) [inline Engine/Module/ModuleAudio.cpp]: why is this commented?

File: Engine/Module/ModuleActions.cpp [hunk 9]
```diff
@@ -130,22 +127,6 @@ void ModuleActions::AddUndoAction(UndoActionType type)
...
-		break; 
-	
-	case UndoActionType::EDIT_RECT2D_ROTATION:
-		new_action = new EditorActionRotation2D(
-			(ComponentTransform2D*)action_component,
-			App->editor->selected_game_object->transform_2d.rotation,
-			App->editor->selected_game_object
-		);
-		break;
-
```
Comment (JorxSS) [inline Engine/Module/ModuleActions.cpp:148]: now the undo/redo cannot longer be applied?

File: Engine/Module/ModuleAudio.cpp (hunk 8 above)
Comment (mariofv) [reply Engine/Module/ModuleAudio.cpp]: This is generating a bug, I'm gonna solve it with Anabel today

File: Engine/Module/ModuleUI.cpp (hunk 2 above)
Comment (mariofv) [reply Engine/Module/ModuleUI.cpp:114]: Nope, just one at the moment. They work similar to cameras

File: Engine/Component/ComponentCanvas.cpp (hunk 5 above)
Comment (mariofv) [reply Engine/Component/ComponentCanvas.cpp:59]: True means that the camera has to take in account that is rendering for the editor `PanelScene`. I use this to render the canvas on the world using a perspective frustum instead of the projection one.

File: Engine/Module/ModuleUI.cpp (hunk 6 above)
Comment (mariofv) [reply Engine/Module/ModuleUI.cpp]: Iep, that's right. Deleting it... 😈

File: Engine/Module/ModuleEditor.cpp (hunk 7 above)
Comment (mariofv) [reply Engine/Module/ModuleEditor.cpp]: I like the idea, implementing it!

File: Engine/Module/ModuleActions.cpp (hunk 9 above)
Comment (mariofv) [reply Engine/Module/ModuleActions.cpp:148]: I delete it because this do/undo was deprecated. As we need to add a lot of new dos/undos, I left the implementation of this for then.

File: Engine/Component/ComponentButton.cpp [hunk 10]
```diff
@@ -5,50 +5,48 @@
...
+void ComponentButton::SpecializedSave(Config& config) const
 {
-	ComponentUI::Save(config);
 }
 
-void ComponentButton::Load(const Config& config)
+void ComponentButton::SpecializedLoad(const Config& config)
 {
-	ComponentUI::Load(config);
 }
```
Comment (Dartemiss) [inline Engine/Component/ComponentButton.cpp]: Why are this functions empty?

File: Engine/Component/ComponentImage.cpp [hunk 11]
```diff
@@ -1,35 +1,129 @@
...
+	{
+		created_component = new ComponentImage();
+	}
+	else
+	{
+		created_component = App->ui->CreateComponentUI<ComponentImage>();
+	}
+	*created_component = *this;
+	return created_component;
+};
```
Comment (Dartemiss) [inline Engine/Component/ComponentImage.cpp:95]: The clone methods are used to duplicate the gameobect and its components. But when duplicating a UI go the transform ignores the cloned object. (this happens with all the clone methods of UI i think)

File: Engine/Filesystem/File.h [hunk 12]
```diff
@@ -11,6 +11,7 @@ enum class FileType
 {
 	ANIMATION,
 	ARCHIVE,
+	FONT,
 	MATERIAL,
 	MESH,
 	META,
```
Comment (Dartemiss) [inline Engine/Filesystem/File.h:17]: Here happens the same thing as with component type right? We have to be careful with audio typefile

**Review by Dartemiss — CHANGES_REQUESTED** [review]
Great job Mario! 
Just pointing out some things taking in account that I'm executing on release:
On this branch gameplay system don't compile as UI has changed the includes needed (now fixed).
Also I strongly suggest to add the option of adding UI by right-clicking on hierarchy as gameobjects.
Also when adding a button it crashes for me

File: Engine/Component/ComponentCanvas.cpp [hunk 13]
```diff
@@ -1,38 +1,127 @@
...
+	for (auto& canvas_renderer : components_to_render)
+	{
+		canvas_renderer->Render(&projection_view);
+	}
+
+	glEnable(GL_DEPTH_TEST);
 }
 
-void ComponentCanvas::Load(const Config& config)
+void ComponentCanvas::SpecializedSave(Config& config) const
```
Comment (JorxSS) [inline Engine/Component/ComponentCanvas.cpp:97]: Specialized Load have logic on it, but the save doesn't, why?

**Review by JorxSS — APPROVED** [review]

**Comment by JorxSS** [comment]
Amazing work, I spent several time with the maths for Transform 2D trying to understand it. Astonoshing.

I'm going to test it further but I cannot add anything else than @Dartemiss .

File: Engine/Component/Component.cpp (hunk 3 above)
Comment (mariofv) [reply Engine/Component/Component.cpp:69]: It was used with the temporal saving of component type. Now it's used when creating UI GameObjects.

File: Engine/Component/Component.h (hunk 4 above)
Comment (mariofv) [reply Engine/Component/Component.h]: thanks!

File: Engine/Component/ComponentButton.cpp (hunk 10 above)
Comment (mariofv) [reply Engine/Component/ComponentButton.cpp]: Because it's a pure virtual function, and in this specific case there is nothing of button to be saved

File: Engine/Filesystem/File.h (hunk 12 above)
Comment (mariofv) [reply Engine/Filesystem/File.h:17]: Here not, because we are not storing anywhere the type as an integer

File: Engine/Component/ComponentCanvas.cpp (hunk 13 above)
Comment (mariofv) [reply Engine/Component/ComponentCanvas.cpp:97]: None of those methods have logic on them 😕

**Review by Dartemiss — APPROVED** [review]
